### ISSUE TITLE
Broker transport failure (errno: Success) ?

### ISSUE THREAD

**Issue by pcullen65** [edenhill/librdkafka#2029]
Read the FAQ first: https://github.com/edenhill/librdkafka/wiki/FAQ

Description
===========
I keep seeing these types of error messages in the producer output:
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,BROKERFAIL]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: failed: err: Local: Broker transport failure: (errno: Succes
s)  [v8.38.0 try http://www.rsyslog.com/e/2422 ]

Any idea what would cause a transport failure, but errno would be Success?

This is only for one broker in a 5-broker cluster. Here's an extract of /var/log/messages around that timeframe for just that broker:

Sep 27 12:13:56 rscope-qa3 rsyslogd: omkafka: kafka log message [7,SEND]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Sent ProduceRequest (v3, 228346 bytes @ 213104, CorrId 11238)  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:13:56 rscope-qa3 rsyslogd: omkafka: kafka log message [7,SEND]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Sent partial ProduceRequest (v3, 0+16496/231223 bytes, CorrId 11239)  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:13:56 rscope-qa3 rsyslogd: omkafka: kafka log message [7,SEND]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Sent partial ProduceRequest (v3, 16496+16384/231223 bytes, CorrId 11239)  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,METADATA]: [thrd:main]: ssl://rscope-dev4.reservoir.com:19093/1:   Broker #0/5: rscope-dev4.reservoir.com:29093 NodeId 2  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,METADATA]: [thrd:main]: ssl://rscope-dev4.reservoir.com:39093/3:   Broker #0/5: rscope-dev4.reservoir.com:29093 NodeId 2  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,BROKERFAIL]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: failed: err: Local: Broker transport failure: (errno: Success)  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,FAIL]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Disconnected  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,STATE]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Broker changed state UP -> DOWN  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,BROADCAST]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: Broadcasting state change  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,BUFQ]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Purging bufq with 0 buffers  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,BUFQ]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Updating 0 buffers on connection reset  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,METADATA]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: Requesting metadata for 1/1 topics: broker down  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,METADATA]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:39093/3: Request metadata for 1 topic(s): broker down  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,METADATA]: [thrd:main]: ssl://rscope-dev4.reservoir.com:39093/3:   Broker #0/5: rscope-dev4.reservoir.com:29093 NodeId 2  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,CONNECT]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: broker in state DOWN connecting  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,CONNECT]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Connecting to ipv4#10.1.0.139:29093 (ssl) with socket 130  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,STATE]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Broker changed state DOWN -> CONNECT  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,BROADCAST]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: Broadcasting state change  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,CONNECT]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Connected to ipv4#10.1.0.139:29093  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,SSLVERIFY]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Broker SSL certificate verified  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,CONNECTED]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Connected (#4)  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,STATE]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Broker changed state CONNECT -> APIVERSION_QUERY  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,BROADCAST]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: Broadcasting state change  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Sep 27 12:14:02 rscope-qa3 rsyslogd: omkafka: kafka log message [7,SEND]: [thrd:ssl://rscope-dev4.reservoir.com:29093/bootstrap]: ssl://rscope-dev4.reservoir.com:29093/2: Sent ApiVersionRequest (v0, 25 bytes @ 0, CorrId 8)  [v8.38.0 try http://www.rsyslog.com/e/2422 ]

The BROKERFAIL occurs about 8s after the last SEND. After the BROKERFAIL, it's able to immediately re-establish the connection.

I've searched the broker logs and there's no indication of a failure during this time.

How to reproduce
================
It's not reproducible, just recurring. This is running rsyslog 8.38 and librdkafka 0.11.5.

**IMPORTANT**: Always try to reproduce the issue on the latest released version (see https://github.com/edenhill/librdkafka/releases), if it can't be reproduced on the latest version the issue has been fixed.

Checklist
=========

**IMPORTANT**: We will close issues where the checklist has not been completed.

Please provide the following information:

 - [x] librdkafka version (release number or git tag): 0.11.5
 - [ ] Apache Kafka version: `<REPLACE with e.g., 0.10.2.3>`
 - [x] librdkafka client configuration: debug=all, statistics.interval.ms=60000
 - [x] Operating system: CentOS 6.5
 - [x] Provide logs (with `debug=..` as necessary) from librdkafka
 - [x] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
Hard to say, the socket was disconnected but no errno was set. OpenSSL does not make it easy to extract the correct error information from an event.
You could try running with `debug=broker,security`

**Comment by pcullen65** [comment]
I was running with debug=all, so I think that should include broker and security, right? In any case, I changed the configuration to remove SSL. Now I see recurring BROKERFAIL messages like this:

Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,BROKERFAIL]: [thrd:rscope-dev4.reservoir.com:49092/4]: rscope-dev4.reservoir.com:49092/4: failed: err: Local: Broker transport failure: (errno: Resource temporarily unavailable)  [v8.38.0 try http://www.rsyslog.com/e/2422 ]

I wonder if SSL was masking the "Resource temporarily unavailable" (EAGAIN) error? It's not much clearer to me what's causing these errors. Again, the broker logs don't show any failures at this time.
Here's a snippet of /var/log/messages for just broker 4 around this time. Is it strange that even while it is in the DOWN state it receives a ProduceResponse?

Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,SEND]: [thrd:rscope-dev4.reservoir.com:49092/4]: rscope-dev4.reservoir.com:49092/4: Sent ProduceRequest (v3, 225575 bytes @ 0, CorrId 12880)  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,TOPPAR]: [thrd:rscope-dev4.reservoir.com:49092/4]: rscope-dev4.reservoir.com:49092/4: conn [9] 1168 message(s) in xmit queue (0 added from partition queue)  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,BROKERFAIL]: [thrd:rscope-dev4.reservoir.com:49092/4]: rscope-dev4.reservoir.com:49092/4: failed: err: Local: Broker transport failure: (errno: Resource temporarily unavailable)  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,FAIL]: [thrd:rscope-dev4.reservoir.com:49092/4]: rscope-dev4.reservoir.com:49092/4: Disconnected  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,STATE]: [thrd:rscope-dev4.reservoir.com:49092/4]: rscope-dev4.reservoir.com:49092/4: Broker changed state UP -> DOWN  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,BROADCAST]: [thrd:rscope-dev4.reservoir.com:49092/4]: Broadcasting state change  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,BUFQ]: [thrd:rscope-dev4.reservoir.com:49092/4]: rscope-dev4.reservoir.com:49092/4: Purging bufq with 0 buffers  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,BUFQ]: [thrd:rscope-dev4.reservoir.com:49092/4]: rscope-dev4.reservoir.com:49092/4: Updating 0 buffers on connection reset  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,METADATA]: [thrd:rscope-dev4.reservoir.com:49092/4]: Requesting metadata for 1/1 topics: broker down  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,METADATA]: [thrd:rscope-dev4.reservoir.com:49092/4]: rscope-dev4.reservoir.com:39092/3: Request metadata for 1 topic(s): broker down  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,RECV]: [thrd:rscope-dev4.reservoir.com:49092/4]: rscope-dev4.reservoir.com:49092/4: Received ProduceResponse (v3, 40 bytes, CorrId 12880, rtt 6.75ms)  [v8.38.0 try http://www.rsyslog.com/e/2422 ]
Oct  1 07:11:18 rscope-qa3 rsyslogd: omkafka: kafka log message [7,MSGSET]: [thrd:rscope-dev4.reservoir.com:49092/4]: rscope-dev4.reservoir.com:49092/4: conn [4]: MessageSet with 1260 message(s) delivered  [v8.38.0 try http://www.rsyslog.com/e/2422 ]

**Comment by edenhill** [comment]
Thank you.
This looks like a network or broker disconnect.
Is there anything in the broker logs?

**Comment by pcullen65** [comment]
It does look like a disconnect, but I've never seen anything in the brokers logs around the time of the disconnect. That's true both with and without SSL.

**Comment by pcullen65** [comment]
I found this was only happening with bootstrap brokers. That made it more likely that the connections were getting reaped as expected, however, the connections should never have been idle, as we were running a soak test with continual log export. I then found that the bootstrap brokers were specified with only the hostname, yet the metadata used the FQDN. That left the bootstrap broker connections in place and going idle; they were reaped every ten minutes, as pointed out here: https://github.com/edenhill/librdkafka/wiki/FAQ#why-am-i-seeing-receive-failed-disconnected
